File: server/src/main/java/com/transparency/service/PackageService.java
package com.transparency.service;

import com.transparency.dao.PackageDAO;
import com.transparency.entity.PackageEntity;
import com.transparency.exception.HierarchyRootNotFoundException;
import com.transparency.exception.PackageNotFoundException;
import com.transparency.model.Package;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class PackageService {

    @Autowired
    private PackageDAO packageDAO;

    public List<Package> findAll() {
        return packageDAO.findAll().stream().map(Package::new).collect(Collectors.toList());
    }

    public Package findById(long id) throws PackageNotFoundException {
        PackageEntity packageEntity = packageDAO.findById(id);
        if (packageEntity == null) {
            throw new PackageNotFoundException(id);
        }
        return new Package(packageEntity);
    }

    public Package findHierarchy() throws HierarchyRootNotFoundException {
        List<Package> allPackages = packageDAO.findAll().stream().map(Package::new).collect(Collectors.toList());
        Package rootPackage = findHierarchyRootIn(allPackages)
                .orElseThrow(HierarchyRootNotFoundException::new);
        return buildHierarchyWithRoot(rootPackage, allPackages);
    }

    private Optional<Package> findHierarchyRootIn(List<Package> packages) {
        return packages.stream().filter(pack -> pack.getParentPackage() == null).findFirst();
    }

    private Package buildHierarchyWithRoot(Package rootPackage, List<Package> packages) {
        packages.remove(rootPackage);
        while (!packages.isEmpty()) {
            int sizeBeforeInserting = packages.size();
            packages = packagesAfterInsertingInto(rootPackage, packages);
            if (packages.size() >= sizeBeforeInserting) {
                break;
            }
        }
        return rootPackage;
    }

    private List<Package> packagesAfterInsertingInto(Package rootPackage, List<Package> packages) {
        Package packageToInsert = packages.get(0);
        if (rootPackage.canAddPackageAsChild(packageToInsert)) {
            rootPackage.addChildPackage(packageToInsert);
        } else {
            packages.add(packageToInsert);
        }
        packages.remove(0);
        return packages;
    }
}

File: server/src/main/kotlin/com/transparency/service/PackageService.kt
package com.transparency.service

import com.transparency.dao.PackageDAO
import com.transparency.exception.HierarchyRootNotFoundException
import com.transparency.exception.PackageNotFoundException
import com.transparency.model.Package
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class PackageService {

    @Autowired
    private lateinit var packageDAO: PackageDAO

    fun findAll(): List<Package> {
        return packageDAO.findAll().map(::Package)
    }

    @Throws(PackageNotFoundException::class)
    fun findById(id: Long): Package {
        val packageEntity = packageDAO.findById(id)
        return Package(packageEntity)
    }

    @Throws(HierarchyRootNotFoundException::class)
    fun findHierarchy(): Package {
        val allPackages = packageDAO.findAllWithFeatureDependencies().map(::Package)
        val rootPackage = findHierarchyRootIn(allPackages) ?: throw HierarchyRootNotFoundException()
        return buildHierarchyWithRoot(rootPackage, allPackages)
    }

    private fun findHierarchyRootIn(packages: List<Package>): Package? {
        return packages.firstOrNull { it.parentPackageId == null }
    }

    private fun buildHierarchyWithRoot(rootPackage: Package, packages: List<Package>): Package {
        var remainingPackages = ArrayList(packages)
        remainingPackages.remove(rootPackage)
        while (!remainingPackages.isEmpty()) {
            val sizeBeforeInserting = packages.size
            remainingPackages = packagesAfterInsertingInto(rootPackage, remainingPackages)
            if (remainingPackages.size >= sizeBeforeInserting) {
                break
            }
        }
        return rootPackage
    }

    private fun packagesAfterInsertingInto(rootPackage: Package, packages: ArrayList<Package>): ArrayList<Package> {
        val packageToInsert = packages[0]
        if (rootPackage.canAddPackageAsChild(packageToInsert)) {
            rootPackage.addChildPackage(packageToInsert)
        } else {
            packages.add(packageToInsert)
        }
        packages.removeAt(0)
        return packages
    }
}
